crates/wasm-encoder/src/functions.rs

```diff
@@ -42,23 +49,43 @@ impl FunctionSection {
         self.num_added == 0
     }
 
-    /// Define a function that uses the given type.
+    /// Define a function in a module's function section.
+    ///
+    /// This method is only supported for encoding modules.
     pub fn function(&mut self, type_index: u32) -> &mut Self {
+        if self.format != SectionEncodingFormat::Module {
+            panic!("functions can only be encoded for modules");
+        }
```

Comment:
I briefly mentioned this above, but I don't think that these panics are necessary and would hope it's ok to remove them. Trying to assert that only a valid wasm module is emitted is I think out of the scope of the `wasm-encoder` crate.

Reply:
I would normally agree, but the intention here is to put the error as close as possible to the problem (e.g. you're using a method on a generic writer type meant for a different format that you specified); relying on `wasmparser` to validate the output means you could get an entirely unhelpful error (e.g. something was incorrectly decoded to a length and now the validator gives an extremely unhelpful "unexpected EOF at byte offset X" error).

Perhaps along with reducing the unnecessary differences in the formats (where possible) the section writers can automatically determine what format they're using based on how they are used and perhaps limit the check to when the section itself gets written to the module/adapter module/component?

Reply:
Some of these checks were removed where possible.